File: src/shared/features/encoders.ts
/**
 * Stable categorical encoders used by the feature-builder.
 *
 * These small tables turn human-readable categorical request fields
 * (`transaction_type`, `transactionChannel`, `idType`, currency code,
 * country code) into small integers that fit a `uint8` ONNX input.
 * Stability matters: changing one of these mappings is a feature-
 * schema change that must come with a catalogue version bump.
 *
 * Unknown values map to 0 — the same slot as "missing" — so an
 * adopter who passes a freshly-added enum value doesn't crash the
 * predict path. The trained model treats 0 as the default for that
 * feature index, identical to the catalogue's `default`.
 */

export function encodeTransactionType(value: string | undefined): number {
  if (!value) return 0;
  switch (value.toUpperCase()) {
    case "CASH_IN":  return 1;
    case "CASH_OUT": return 2;
    case "PAYMENT":  return 3;
    case "TRANSFER": return 4;
    case "DEBIT":    return 5;
    case "PAYOUT":   return 6;
    case "WITHDRAWAL": return 7;
    default: return 0;
  }
}

export function encodeChannel(value: string | undefined): number {
  if (!value) return 0;
  switch (value.toUpperCase()) {
    case "USSD":           return 1;
    case "MOBILE":         return 2;
    case "WEB":            return 3;
    case "AGENT":          return 4;
    case "POS":            return 5;
    case "ATM":            return 6;
    case "API":            return 7;
    default: return 0;
  }
}

/**
 * Coarse currency encoding — ISO-4217 numeric collapsed to small
 * ints so the value fits a uint8. Add more entries here when an
 * adopter needs them; bump the catalogue version when you do.
 */
export function encodeCurrency(value: string | undefined): number {
  if (!value) return 0;
  switch (value.toUpperCase()) {
    case "NGN": return 1;
    case "USD": return 2;
    case "EUR": return 3;
    case "GBP": return 4;
    case "ZAR": return 5;
    case "KES": return 6;
    case "GHS": return 7;
    default: return 0;
  }
}

export function encodeIdType(value: string | undefined): number {
  if (!value) return 0;
  switch (value.toUpperCase()) {
    case "BVN":              return 1;
    case "NIN":              return 2;
    case "PASSPORT":         return 3;
    case "DRIVERS_LICENSE":  return 4;
    case "NATIONAL_ID":      return 5;
    case "VOTERS_CARD":      return 6;
    case "OTHER_ID":         return 7;
    default: return 0;
  }
}

/**
 * Encode a country to a uint8. Uses ISO-3166 alpha-2 input and
 * collapses to a small numeric. NG is first because it's the
 * dominant nationality in this codebase's expected workload; the
 * rest follow a stable alphabetical ordering for the African
 * countries we expect, then "OTHER" for everything else.
 */
const COUNTRY_TO_CODE: ReadonlyMap<string, number> = new Map([
  ["NG", 1], ["GH", 2], ["KE", 3], ["ZA", 4], ["EG", 5],
  ["US", 6], ["GB", 7], ["DE", 8], ["FR", 9], ["IN", 10],
  ["CN", 11], ["AE", 12],
]);

export function encodeCountry(value: string | undefined): number {
  if (!value) return 0;
  return COUNTRY_TO_CODE.get(value.toUpperCase()) ?? 255; // 255 = OTHER
}

/**
 * Encode device type. Same convention as channel.
 */
export function encodeDeviceType(value: string | undefined): number {
  if (!value) return 0;
  switch (value.toUpperCase()) {
    case "MOBILE":          return 1;
    case "WEB":             return 2;
    case "USSD":            return 3;
    case "POS":             return 4;
    case "AGENT_TERMINAL":  return 5;
    case "ATM":             return 6;
    default: return 0;
  }
}

/**
 * Common categorical/scalar tools used by compute ops.
 */
export function safeNumber(value: unknown, fallback = 0): number {
  if (value == null) return fallback;
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string") {
    const n = Number(value);
    return Number.isFinite(n) ? n : fallback;
  }
  if (typeof value === "boolean") return value ? 1 : 0;
  return fallback;
}

export function safeBool(value: unknown): 0 | 1 {
  if (value == null) return 0;
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value === "number") return value > 0 ? 1 : 0;
  if (typeof value === "string") {
    const v = value.trim().toLowerCase();
    return v === "true" || v === "yes" || v === "y" || v === "1" ? 1 : 0;
  }
  return 0;
}

/**
 * Haversine distance in km between two (lat, lng) pairs.
 * Returns 0 when either coordinate pair is missing or invalid.
 */
export function haversineKm(
  lat1: number | undefined,
  lng1: number | undefined,
  lat2: number | undefined,
  lng2: number | undefined
): number {
  if (lat1 == null || lng1 == null || lat2 == null || lng2 == null) return 0;
  if (
    !Number.isFinite(lat1) ||
    !Number.isFinite(lng1) ||
    !Number.isFinite(lat2) ||
    !Number.isFinite(lng2)
  ) {
    return 0;
  }
  const R = 6371;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(a));
}

/**
 * Whole days between an ISO date string and `now`. Returns 0 on
 * unparseable / missing input — the catalogue defines that as the
 * "unknown" slot for *_age_days features.
 */
export function ageDays(iso: string | undefined, now = Date.now()): number {
  if (!iso) return 0;
  const t = Date.parse(iso);
  if (!Number.isFinite(t)) return 0;
  const days = Math.floor((now - t) / (1000 * 60 * 60 * 24));
  return Math.max(0, days);
}

File: mla-service/src/features/encoders.py
"""
Python mirror of `src/shared/features/encoders.ts`.

These tables MUST stay in lock-step with the TS encoders — any change
here without the matching TS edit (or vice versa) creates a silent
train/serve skew. The `feature_schema_version` enforcement at RDA load
time is the loud-failure mechanism; this module is just the Python
half of the contract.

Vectorised wrappers (`encode_*_series`) operate on pandas Series so the
training data_loader can build columns in one shot.
"""

from __future__ import annotations

import math
from datetime import datetime, timezone
from typing import Optional

import numpy as np
import pandas as pd


# ── Categorical encoders (mirror TS exactly) ────────────────────

_TXN_TYPE = {
    "": 0, "CASH_IN": 1, "CASH_OUT": 2, "PAYMENT": 3, "TRANSFER": 4,
    "DEBIT": 5, "PAYOUT": 6, "WITHDRAWAL": 7,
}

_CHANNEL = {
    "": 0, "USSD": 1, "MOBILE": 2, "WEB": 3, "AGENT": 4,
    "POS": 5, "ATM": 6, "API": 7,
}

_CURRENCY = {
    "": 0, "NGN": 1, "USD": 2, "EUR": 3, "GBP": 4,
    "ZAR": 5, "KES": 6, "GHS": 7,
}

_ID_TYPE = {
    "": 0, "BVN": 1, "NIN": 2, "PASSPORT": 3, "DRIVERS_LICENSE": 4,
    "NATIONAL_ID": 5, "VOTERS_CARD": 6, "OTHER_ID": 7,
}

_COUNTRY = {
    "NG": 1, "GH": 2, "KE": 3, "ZA": 4, "EG": 5,
    "US": 6, "GB": 7, "DE": 8, "FR": 9, "IN": 10,
    "CN": 11, "AE": 12,
}

_DEVICE_TYPE = {
    "": 0, "MOBILE": 1, "WEB": 2, "USSD": 3, "POS": 4,
    "AGENT_TERMINAL": 5, "ATM": 6,
}


def _encode_series(series: pd.Series, table: dict, unknown: int = 0) -> pd.Series:
    """Upper-case, strip, map through `table`. Unknowns → `unknown`."""
    return (
        series.fillna("")
        .astype(str)
        .str.upper()
        .str.strip()
        .map(table)
        .fillna(unknown)
    )


def encode_transaction_type(series: pd.Series) -> pd.Series:
    return _encode_series(series, _TXN_TYPE, unknown=0)


def encode_channel(series: pd.Series) -> pd.Series:
    return _encode_series(series, _CHANNEL, unknown=0)


def encode_currency(series: pd.Series) -> pd.Series:
    return _encode_series(series, _CURRENCY, unknown=0)


def encode_id_type(series: pd.Series) -> pd.Series:
    return _encode_series(series, _ID_TYPE, unknown=0)


def encode_device_type(series: pd.Series) -> pd.Series:
    return _encode_series(series, _DEVICE_TYPE, unknown=0)


def encode_country(series: pd.Series) -> pd.Series:
    """Unknowns map to 255 (TS reserves 0 for missing, 255 for OTHER)."""
    upper = series.fillna("").astype(str).str.upper().str.strip()
    # Missing → 0; recognised → code; unrecognised non-empty → 255.
    out = upper.map(_COUNTRY)
    out = out.where(upper != "", 0)
    return out.fillna(255).astype("float32")


# ── Numeric helpers ─────────────────────────────────────────────


def safe_number_series(series: pd.Series, fallback: float = 0.0) -> pd.Series:
    return pd.to_numeric(series, errors="coerce").fillna(fallback)


def safe_bool_series(series: pd.Series) -> pd.Series:
    """Coerce arbitrary booleans / 0/1 / 'yes'/'no' strings to {0, 1}."""
    if series.dtype == bool:
        return series.astype("int8")
    truthy = series.fillna("").astype(str).str.strip().str.lower()
    return truthy.isin(["true", "yes", "y", "1"]).astype("int8")


def haversine_km(
    lat1: pd.Series, lng1: pd.Series, lat2: pd.Series, lng2: pd.Series
) -> pd.Series:
    """Vectorised great-circle distance in km. Missing inputs → 0."""
    a = safe_number_series(lat1, np.nan)
    b = safe_number_series(lng1, np.nan)
    c = safe_number_series(lat2, np.nan)
    d = safe_number_series(lng2, np.nan)

    mask = a.notna() & b.notna() & c.notna() & d.notna()
    if not mask.any():
        return pd.Series(np.zeros(len(a)), index=a.index)

    R = 6371.0
    lat1r = np.radians(a)
    lat2r = np.radians(c)
    dlat = np.radians(c - a)
    dlng = np.radians(d - b)
    h = (
        np.sin(dlat / 2.0) ** 2
        + np.cos(lat1r) * np.cos(lat2r) * np.sin(dlng / 2.0) ** 2
    )
    out = 2.0 * R * np.arcsin(np.sqrt(np.clip(h, 0, 1)))
    return pd.Series(out, index=a.index).where(mask, 0.0)


def age_days_series(dob_series: pd.Series, now_ts: float) -> pd.Series:
    """Days between an ISO/date column and `now_ts` (epoch seconds)."""
    try:
        parsed = pd.to_datetime(dob_series, errors="coerce", utc=True)
    except Exception:
        return pd.Series(np.zeros(len(dob_series)), index=dob_series.index)
    now = datetime.fromtimestamp(now_ts, tz=timezone.utc)
    diff = (now - parsed).dt.total_seconds() / 86400.0
    return diff.fillna(0).clip(lower=0)
